torch/lib/c10d/ProcessGroupGloo.hpp

```diff
@@ -251,8 +251,13 @@ class ProcessGroupGloo : public ProcessGroup {
   c10::intrusive_ptr<ProcessGroup::Work> barrier(
       const BarrierOptions& opts = BarrierOptions()) override;
 
+  // Similar to barrier(), but blocks rank 0 until all other ranks have
+  // acknowledged that they are alive (through send/recv from rank 0). Rank 0
+  // is able to report all failed ranks if waitAllRanks = true, otherwise
+  // reports the first rank it detected as failed.
   void monitoredBarrier(
-      const BarrierOptions& opts = BarrierOptions()) override;
+      const BarrierOptions& opts = BarrierOptions(),
+      bool waitAllRanks = false) override;
```

Comment:
I don't see an easy way to flip this in practice. What about automatically turning on this whenever we use a debug mode (e.g., `NCCL_DEBUG=INFO`)

Reply:
This PR exposes this flag to the python interface, so users can just do `dist.monitored_barrier(wait_all_ranks=True)`. Although agreed, it is important to think about the default since it may be more useful to collect _all_ ranks if user is using this debug tool. Since this is mostly going to be used in conjunction with torch_distributed_debug, maybe wait_all_ranks=True by default makes sense, though wait_all_ranks=False provides a "fail fast" option.

Reply:
My bad, I didn't realize it wasn't exposed to python in the version you reviewed. Added that now.

Reply:
> This PR exposes this flag to the python interface, so users can just do `dist.monitored_barrier(wait_all_ranks=True)`. Although agreed, it is important to think about the default since it may be more useful to collect _all_ ranks if user is using this debug tool. Since this is mostly going to be used in conjunction with torch_distributed_debug, maybe wait_all_ranks=True by default makes sense, though wait_all_ranks=False provides a "fail fast" option.

We don't know what kind of error the user is going to investigate, aka., whether collecting all the ranks here will help. You can make a decision here based on which case is more common.

Reply:
Good point, will land this change and spend some time thinking about the default after collecting some user feedback.